django/restapi/views.py

```diff
@@ -67,54 +75,297 @@ def post(self, request, format=None):
+def getUser(request):
+
+    serializer = UserSerializer(request.user)
+    return JsonResponse(serializer.data, safe=False)
+
+
+# /companies
+@api_view(['GET'])
+def getCompanies(request):
+
+    # Check if authenticated user is allowed to request company_id, garden_id
+
+    try:
+        companies = Company.objects.filter(user=request.user)
+    except:
+        # Company with id company_id not found in database or does not belong to the user requesting it
+        return HttpResponseNotFound()
+
+    serializer = CompanySerializer(companies, many=True)
+    return JsonResponse(serializer.data, safe=False)
 
 
 # /companies/{company_id}
 @api_view(['GET'])
 def getCompany(request, company_id: int):
-    company = Company.objects.filter(id=company_id)
-    serializer = CompanySerializer(company, many=True)
+
+    # Check if authenticated user is allowed to request company_id, garden_id
+
+    try:
+        company = Company.objects.get(id=company_id, user=request.user)
+    except:
+        # Company with id company_id not found in database or does not belong to the user requesting it
+        return HttpResponseNotFound()
+
+    serializer = CompanySerializer(company)
     return JsonResponse(serializer.data, safe=False)
 
 
 # /companies/{company_id}/gardens
 @api_view(['GET'])
 def getGardens(request, company_id: int):
-    garden = Garden.objects.filter(company=company_id)
-    serializer = GardenSerializer(garden, many=True)
+
+    # Check if authenticated user is allowed to request company_id, garden_id
+
+    try:
+        company = Company.objects.get(id=company_id, user=request.user)
+    except:
+        # Company with id company_id not found in database or does not belong to the user requesting it
+        return HttpResponseNotFound()
+
+    gardens = Garden.objects.filter(company=company)
+    serializer = GardenSerializer(gardens, many=True)
     return JsonResponse(serializer.data, safe=False)
 
 
 # /companies/{company_id}/gardens/{garden_id}
 @api_view(['GET'])
 def getGarden(request, company_id: int, garden_id: int):
-    garden = Garden.objects.filter(id=garden_id, company=company_id)
-    serializer = GardenSerializer(garden, many=True)
+
+    # Check if authenticated user is allowed to request company_id, garden_id
+
+    try:
+        company = Company.objects.get(id=company_id, user=request.user)
+    except:
+        # Company with id company_id not found in database or does not belong to the user requesting it
+        return HttpResponseNotFound()
+
+    try:
+        garden = Garden.objects.get(id=garden_id, company=company)
+    except:
+        # Garden with id garden_id not found in database or does not belong to the company
+        return HttpResponseNotFound()
+
+    serializer = GardenSerializer(garden)
     return JsonResponse(serializer.data, safe=False)
 
 
 # /companies/{company_id}/gardens/{garden_id}/image
 @api_view(['GET', 'POST'])
 def gardenImage(request, company_id: int, garden_id: int):
```

Comment:
Wouldn't it be a cleaner approach to split the request explicitly into GET and POST? Then you don't need to check the corresponding HTTP method.

Reply:
😢